File: MATLAB Intro/GraphsandTimeSeries.m
%Code by Javier Gonzalez on October 2019
%For any questions please contact javierj.g18@gmail.com
% Code intended to teach graphing tools, for loops and if statements
% Some stochastic processes

clear all
%Setiamos la cantidad de periodos a ver
T = 100;
%Alocamos memoria
MA1 = zeros(T,1);
MA2 = zeros(T,1);
AR = zeros(T,1);
ARMA = zeros(T,1);

% Elegimos un numero random para empezar
MA1(1) = rand;
MA2(1) = rand;
AR(1) = rand;
ARMA(1) = rand;
% Matrices de errores
Error =rand(T,1);

%Comenzando con un MA(1)
%Definimos el coeficiente
alpha = 1.005; % Lo defino de esta forma para poder usarlo como estacionario
for i = 2:T % Ojo que comienzo en 2 y no en 1
    MA1(i)= alpha*Error(i-1) + Error(i);
end

%Comenzando con un MA(2)
%Necesitamos un segundo coeficiente y sabemos que si comienzo en 2 tendre
%un problema, pues no existe un subindice 0, tenemos que poner un if.
beta = 0.34; % Lo defino de esta forma para poder usarlo como estacionario
for i = 2:T % Ojo que comienzo en 2 y no en 1
    if i == 2
        MA2(i)= alpha*Error(i-1) + Error(i);
    else
        MA2(i)= beta*Error(i-2) + alpha*Error(i-1) + Error(i);
    end
end

%Comenzando con un AR(1) estacionario
%Como tenemos un coeficiente que es menor a 1 (beta) lo usaremos
for i = 2:T % Ojo que comienzo en 2 y no en 1
    AR(i)= beta*AR(i-1) + Error(i);
end

%Comenzando con un ARMA(1,2) no estacionario
%Necesitamos un tercer coeficiente
%voy a cambiar un poco y hacer esto como un while loop
sigma = 0.7; % Lo defino de esta forma para poder usarlo como estacionario
t=2;
while t <= T
    if t == 2
        ARMA(t)= alpha*ARMA(t-1) + 0 + sigma*Error(t-1) + Error(t);
    else
        ARMA(t)= alpha*ARMA(t-1) + beta*Error(t-2) + sigma*Error(t-1) + Error(t);
    end
    t = t+1;
end

% Ya que tenemos todas las series veamos un grafico de 2x2 para ver cada
% una.
% necesitamos setiar un gridspace
grid = linspace(0,T);
%Aqui graficamos ambas economias
figure
subplot(2,2,1)
plot(grid,MA1)
title('MA1')
xlabel('Tiempo')
subplot(2,2,2)
plot(grid,MA2)
title('MA2')
xlabel('Tiempo')
subplot(2,2,3)
plot(grid,AR)
title('AR1 Estacionario')
xlabel('Tiempo')
subplot(2,2,4)
plot(grid,ARMA)
title('ARMA(1,2) NO Estacionario')
xlabel('Tiempo')

File: MATLAB Intro/GraphsandTImeSeries.py
#Code by Javier Gonzalez on December 2019
#For any questions please contact javierj.g18@gmail.com
# Code intended to teach graphing tools, for loops and if statements
# Some stochastic processes

import numpy as np
import matplotlib.pyplot as plt

np.random.seed(20)
def MA1(et,μ=1.005):
    return μ*et

def MA2(et,et_1,μ=1.005,α=0.34):
    return μ*et+α*et_1

def AR1(xt_1,et,β=0.34):
    return xt_1*β+et

def ARMA(xt_1,et_2,et_1,et,γ=0.34,β=1.005,sigma=0.7):
    return xt_1*β+et_1*γ+et_2*sigma+et

n = 100
e = abs(np.random.randn(n))
seriesMA1 = []
seriesMA2 = []
seriesAR1 = []
seriesARMA = []
zero = []
for t in range(n):
    zero.append(np.mean(e))
    a1 = MA1(e[t])
    seriesMA1.append(a1)
    if t == 0:
        a2 = AR1(xt_1 = 0, et=  e[t])
        seriesAR1.append(a2)
        a3 = MA2(e[t],0)
        seriesMA2.append(a3)
        a4 = ARMA(xt_1 = 0 , et_2=0, et_1=0,et=  e[t])
        seriesARMA.append(a4)
    elif t == 1:
        a2 = AR1(xt_1 = seriesAR1[t-1] , et=  e[t])
        seriesAR1.append(a2)
        a3 = MA2(e[t],e[t-1])
        seriesMA2.append(a3)
        a4 = ARMA(xt_1 = seriesARMA[t-1] , et_2=0, et_1=e[t-1],et=  e[t])
        seriesARMA.append(a4)
    else:
        a2 = AR1(xt_1 = seriesAR1[t-1] , et=  e[t])
        seriesAR1.append(a2)
        a3 = MA2(e[t],e[t-1])
        seriesMA2.append(a3)
        a4 = ARMA(xt_1 = seriesARMA[t-1] , et_2=e[t-2], et_1=e[t-1],et=  e[t])
        seriesARMA.append(a4)

fig, axes = plt.subplots(2, 2, figsize=(10, 12))
axes[0, 0].plot(seriesMA1)
axes[0, 1].plot(seriesMA2)
axes[1, 0].plot(seriesAR1)
axes[1, 1].plot(seriesARMA)
# Esto es para poner una linea en torno a la media del error
axes[0, 0].plot(zero,'--k',alpha=0.2)
axes[0, 1].plot(zero,'--k',alpha=0.2)
axes[1, 0].plot(zero,'--k',alpha=0.2)
axes[1, 1].plot(zero,'--k',alpha=0.2)
# Titulos
axes[0, 0].set(title="MA(1)")
axes[0, 1].set(title="MA(2)")
axes[1, 0].set(title="AR(1)")
axes[1, 1].set(title="ARMA(1,2)")
plt.show()